core/src/main/scala/kafka/log/Log.scala

```diff
@@ -403,9 +456,31 @@ class Log(@volatile var dir: File,
         time = time,
         fileSuffix = SwapFileSuffix)
       info(s"Found log file ${swapFile.getPath} from interrupted swap operation, repairing.")
-      recoverSegment(swapSegment)
-      val oldSegments = logSegments(swapSegment.baseOffset, swapSegment.readNextOffset)
-      replaceSegments(swapSegment, oldSegments.toSeq, isRecoveredSwapFile = true)
+
+      try recoverSegment(swapSegment)
+      catch {
+        case e: IndexOffsetOverflowException => {
+          error(s"Unexpected IndexOffsetOverflowException for swap segment ${swapSegment.log.file.getAbsolutePath}")
+          throw e
```

Comment:
I think this will propagate all the way up to `KafkaServer#startup` and will be treated as a fatal error that shuts down the broker. Thinking about this more, we might want to split swap segments with offset overflow as well but we probably don't have a good way to do this given the current code. As discussed offline, because this is expected to be a rare scenario, I will document this situation and leave it as is for now.